### PR TITLE
Refactor ManagedWebSocket to avoid forcing Task allocations for ReceiveAsync

### LINKED ISSUE #50921: Use the new async state machine pooling feature on websockets
### Description

Looking at an allocation profile for SignalR, the top allocations are coming from state machines allocated by Kestrel on websocket reads. We should be able to remove these by using the new pooling introduced here https://github.com/dotnet/runtime/pull/50116.

### Regression?

No

### Data

image

### Analysis

This is the code path:

https://github.com/dotnet/runtime/blob/d927fcbb36131b8962d77d9f01bb3c6112072b95/src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs#L685

cc @stephentoub @CarnaViire

### PR DESCRIPTION
**Pull request by stephentoub**
The ManagedWebSocket implementation today supports CloseAsyncs being issued concurrently with ReceiveAsyncs, even though CloseAsync needs to issue receives (this allowance was carried over from the .NET Framework implementation).  Currently the implementation does that by storing the last ReceiveAsync task and awaiting it in CloseAsync if there is one, but that means multiple parties may try to await the same task multiple times (the original caller of ReceiveAsync and CloseAsync), which means we can't just use a ValueTask.  So today asynchronously completing ReceiveAsyncs always use AsTask to create a Task from the returned ValueTask.  This isn't actually an additional task allocation today, as the async ValueTask builder will create a Task for the asynchronously completing operation, and then AsTask will just return that (and when it completes synchronously, there's extra code to substitute a singleton).  But once we switch to using the new pooling builder, that's no longer the case.

This PR uses an async lock as part of the ReceiveAsync implementation, with the existing async method awaiting entering that lock.  CloseAsync is then rewritten to be in terms of calling ReceiveAsync in a loop.  This also lets us remove the existing Monitor used for synchronously coordinating state between these operations, as the async lock serves that purpose as well.  Rather than using a SemaphoreSlim, since we expect zero contention in the common case, we use a simple AsyncMutex that's optimized for the zero contention case, using a single interlocked to acquire and a single interlocked to release the lock.

Closes https://github.com/dotnet/runtime/issues/50921

|   Method |         Toolchain |     Mean |   Error |  StdDev | Ratio |      Gen 0 |     Gen 1 |    Gen 2 |  Allocated |
|--------- |------------------ |---------:|--------:|--------:|------:|-----------:|----------:|---------:|-----------:|
| PingPong | \main\corerun.exe | 148.7 ms | 2.92 ms | 4.00 ms |  1.00 | 29750.0000 | 3000.0000 | 250.0000 | 180,238 KB |
| PingPong |   \pr\corerun.exe | 108.9 ms | 1.56 ms | 1.38 ms |  0.72 |          - |         - |        - |     249 KB |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Net.WebSockets;

[MemoryDiagnoser]
public class Program
{
    public static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private class Connection
    {
        public readonly WebSocket Client, Server;
        public readonly Memory<byte> ClientBuffer = new byte[256];
        public readonly Memory<byte> ServerBuffer = new byte[256];
        public readonly CancellationToken CancellationToken = default;

        public Connection()
        {
            (Stream Stream1, Stream Stream2) streams = ConnectedStreams.CreateBidirectional();
            Client = WebSocket.CreateFromStream(streams.Stream1, isServer: false, subProtocol: null, Timeout.InfiniteTimeSpan);
            Server = WebSocket.CreateFromStream(streams.Stream2, isServer: true, subProtocol: null, Timeout.InfiniteTimeSpan);
        }
    }

    private Connection[] _connections = Enumerable.Range(0, 256).Select(_ => new Connection()).ToArray();
    private const int Iters = 1_000;

    [Benchmark]
    public Task PingPong() =>
        Task.WhenAll(from c in _connections select Task.WhenAll(
                         Task.Run(async () =>
                         {
                             for (int i = 0; i < Iters; i++)
                             {
                                 await c.Server.ReceiveAsync(c.ServerBuffer, c.CancellationToken);
                                 await c.Server.SendAsync(c.ServerBuffer, WebSocketMessageType.Binary, endOfMessage: true, c.CancellationToken);
                             }
                         }),
                         Task.Run(async () =>
                         {
                             for (int i = 0; i < Iters; i++)
                             {
                                 await c.Client.SendAsync(c.ClientBuffer, WebSocketMessageType.Binary, endOfMessage: true, c.CancellationToken);
                                 await c.Client.ReceiveAsync(c.ClientBuffer, c.CancellationToken);
                             }
                         })));
}
```

### CODE REVIEW

**Comment by stephentoub** [comment]
@davidfowl, can you help validate this with ASP.NET functional tests and against relevant ASP.NET perf tests? That needs to be done before  this can be merged.

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/AsyncMutex.cs
Comment (stephentoub) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/AsyncMutex.cs:15]: Note this is just putting back the AsyncMutex that was removed (due to no longer being necessary for the use case) in https://github.com/dotnet/runtime/pull/39166/files#diff-bd308ac282a26fd004ac30d3c41eee3d90338f4b7a1c7bdfdb8b8488345490e5L107.
The only change from that is I simplified the Waiter type to just be a TaskCompletionSource, and removed all of the pooling that was being done, since we don't expect contention here that warrants that additional complexity.

**Review by karelz — APPROVED** [review]
@CarnaViire if you get a chance to take a look as well, that would be good.

**Comment by stephentoub** [comment]
@davidfowl, @adityamandaleeka, any update on validating this change for ASP.NET?

**Review by CarnaViire — APPROVED** [review]
LGTM

**Comment by stephentoub** [comment]
Once CI is green, I'll go ahead and merge this.  In my own tests, this shows up as neutral to positive both locally in microbenchmarks and on asp-perf-lin and asp-citrine-lin in terms of throughput.  We can revert it if it ends up having any negative impact once it makes it to dotnet/aspnetcore.

@davidfowl, @adityamandaleeka, I'd still appreciate extra validation here, but at this point if I don't merge it's not going to make the release.  The new websockets benchmark doesn't seem to really stress the system with or without this.

**Comment by davidfowl** [comment]
Merging and propagating a dependency flow PR is the easiest way to get validation.

**Pull request merged**
